### ISSUE TITLE
PCH step doesn't depend on generated headers

### ISSUE THREAD

**Issue by nicolas17** [mesonbuild/meson#2361]
If a header to be precompiled includes a header that is generated during the build, meson doesn't create a proper dependency between them. This leads to ninja trying to precompile the pch before the headers it includes are generated.

Full example:

main.c:
```c
#include <stdio.h>
#include "bigheader.h"

int main() {
    printf("The answer is %d\n", x);
    return 0;
}
```

generator.sh:
```sh
#!/bin/sh
sleep 2
echo "#pragma once" > $1
echo "const int x = 42;" >> $1
```

pch/pch.h:
```c
#include "bigheader.h"
```

meson.build:
```python
project('bugtest', 'c')

# pretend this generates a giant file
gen = custom_target('header',
    output: ['bigheader.h'],
    command: ['sh', files('generator.sh'), '@OUTPUT@']
)

executable('bugtest', ['main.c', gen], c_pch: 'pch/pch.h')
```

The resulting build.ninja has an order-only dependency from `main.c.o` to `pch.h.gch bigheader.h` which ensures the compilation waits for `bigheader.h` to be generated, but there is nothing making `bigheader.h` a dependency of the PCH step.

On a `ninja -j1` this *might* work, but in a parallel build this leads to pch.h being precompiled before bigheader.h is generated.

One way to solve this is to have meson put all the target's generated headers as order-only dependencies of the precompiled header. Although this might cause too-strict dependencies for the common case where the pch only includes external libraries, in which case I guess this needs an extra keyword argument to be explicit...

**Comment by nioncode** [comment]
Another way would be to have PCH as a separate pseudo-target or just a method that can take source + header arguments, as well as a list of dependencies. Then, we could add arbitrary dependencies before generating the PCH without adding ALL generators as dependencies.

**Comment by jpakkane** [comment]
The dependencies of pch compilation should be identical to regular compilation. It is not because of historical reasons.

**Comment by nioncode** [comment]
Fixed by #4860, please close.
